impeller/entity/shaders/blending/porter_duff_blend.frag

```diff
@@ -0,0 +1,74 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <impeller/blending.glsl>
+#include <impeller/color.glsl>
+#include <impeller/texture.glsl>
+#include <impeller/types.glsl>
+
+uniform f16sampler2D texture_sampler_dst;
+
+// This must be kept in sync with the values in impeller/geometry/color.h
+// clear, src, dst are excluded as these do not require blending.
+
+const float16_t kSourceOver = 3.0hf;
+const float16_t kDestinationOver = 4.0hf;
+const float16_t kSourceIn = 5.0hf;
+const float16_t kDestinationIn = 6.0hf;
+const float16_t kSourceOut = 7.0hf;
+const float16_t kDestinationOut = 8.0hf;
+const float16_t kSourceATop = 9.0hf;
+const float16_t kDestinationATop = 10.0hf;
+const float16_t kXor = 11.0hf;
+const float16_t kPlus = 12.0hf;
+const float16_t kModulate = 13.0hf;
+
+uniform FragInfo {
+  float16_t operation;
+  float16_t input_alpha;
+  f16vec4 color;
+}
+frag_info;
+
+in vec2 v_texture_coords;
+
+out f16vec4 frag_color;
+
+f16vec4 Sample(f16sampler2D texture_sampler, vec2 texture_coords) {
+// gles 2.0 is the only backend without native decal support.
+#ifdef IMPELLER_TARGET_OPENGLES
+  return IPSampleDecal(texture_sampler, texture_coords);
+#else
+  return texture(texture_sampler, texture_coords);
+#endif
+}
+
+void main() {
+  f16vec4 dst_color =
+      texture(texture_sampler_dst, v_texture_coords) * frag_info.input_alpha;
+
+  if (frag_info.operation == kSourceOver) {
+    frag_color = IPBlendSourceOver(frag_info.color, dst_color);
+  } else if (frag_info.operation == kDestinationOver) {
+    frag_color = IPBlendDestinationOver(frag_info.color, dst_color);
+  } else if (frag_info.operation == kSourceIn) {
+    frag_color = IPBlendSourceIn(frag_info.color, dst_color);
+  } else if (frag_info.operation == kDestinationIn) {
+    frag_color = IPBlendDestinationIn(frag_info.color, dst_color);
+  } else if (frag_info.operation == kSourceOut) {
+    frag_color = IPBlendSourceOut(frag_info.color, dst_color);
+  } else if (frag_info.operation == kDestinationOut) {
+    frag_color = IPBlendDestinationOut(frag_info.color, dst_color);
+  } else if (frag_info.operation == kSourceATop) {
+    frag_color = IPBlendSourceATop(frag_info.color, dst_color);
+  } else if (frag_info.operation == kDestinationATop) {
+    frag_color = IPBlendDestinationATop(frag_info.color, dst_color);
+  } else if (frag_info.operation == kXor) {
+    frag_color = IPBlendXOR(frag_info.color, dst_color);
+  } else if (frag_info.operation == kPlus) {
+    frag_color = IPBlendPlus(frag_info.color, dst_color);
+  } else {
+    frag_color = IPBlendModulate(frag_info.color, dst_color);
+  }
```

Comment:
Rather than enumerating these on the GPU, most of these can be parameterized with 4 coefficients:
```glsl
frag_color =
    src * (frag_info.src_coeff + dst.a * frag_info.src_coeff_dst_alpha) +
    dst * (frag_info.dst_coeff + src.a * frag_info.dst_coeff_src_alpha);
```

And then the coefficients can be pulled from a LUT on the CPU without enumeration or hashing:

```cpp
{src_coeff, src_coeff_dst_alpha, dst_coeff, dst_coeff_src_alpha}

{0,  0,  0,  0}, // Clear
{1,  0,  0,  0}, // Source
{0,  0,  1,  0}, // Destination
{1,  0,  1, -1}, // SourceOver
{1, -1,  1,  0}, // DestinationOver
{0,  1,  0,  0}, // SourceIn
{0,  0,  0,  1}, // DestinationIn
{1, -1,  0,  0}, // SourceOut
{0,  0,  1, -1}, // DestinationOut
{0,  1,  1, -1}, // SourceATop
{1, -1,  0,  1}, // DestinationATop
{1, -1,  1, -1}, // Xor
{1,  0,  1,  0}, // Plus
```

I think we can also cover `Modulate` and `Screen` here by adding one more dst coefficient to multiply the source color in:

```glsl
frag_color =
    src * (frag_info.src_coeff + dst.a * frag_info.src_coeff_dst_alpha) +
    dst * (frag_info.dst_coeff + src.a * frag_info.dst_coeff_src_alpha +
                                 src * frag_info.dst_coeff_src_color);
```

...and then the full table with all the blends becomes:

```cpp
{src_coeff, src_coeff_dst_alpha, dst_coeff, dst_coeff_src_alpha, dst_coeff_src_color}

{0,  0,  0,  0,  0}, // Clear
{1,  0,  0,  0,  0}, // Source
{0,  0,  1,  0,  0}, // Destination
{1,  0,  1, -1,  0}, // SourceOver
{1, -1,  1,  0,  0}, // DestinationOver
{0,  1,  0,  0,  0}, // SourceIn
{0,  0,  0,  1,  0}, // DestinationIn
{1, -1,  0,  0,  0}, // SourceOut
{0,  0,  1, -1,  0}, // DestinationOut
{0,  1,  1, -1,  0}, // SourceATop
{1, -1,  0,  1,  0}, // DestinationATop
{1, -1,  1, -1,  0}, // Xor
{1,  0,  1,  0,  0}, // Plus
{0,  0,  0,  0,  1}, // Modulate
{0,  0,  1,  0, -1}, // Screen
```

+cc @flar in case he has a moment to check my math :)

Reply:
Seems like a more technically sound approach, I'll make this change.

Reply:
Done, lines up with the current behavior using a 0.5 alpha red and blue:

ToT

![flutter_01](https://user-images.githubusercontent.com/8975114/232175293-30937c8f-a910-4290-aa61-b9e5f1dcce84.png)

Patched:

![flutter_03](https://user-images.githubusercontent.com/8975114/232175279-b9644393-f9b6-4e6d-9b8b-0fd49c61b0bf.png)